### PR TITLE
[FIX] tools: classify \uFEFF as whitespace for translations

### PR DESCRIPTION
**Pull request by lvsz**
Versions
--------
- 15.0
- 16.0
- 17.0
- saas-17.1
- saas-17.2
- saas-17.3

Issue
-----
  Commit 9426ee54b927 introduced the \uFEFF character to web_editor as a zero-width non-breaking whitespace. When this gets added to a HTML node, and processed for translation, it throws an "empty document" error.

Cause
-----
When passed to the `get_text_content` function, the call to `html.fromstring('\uFEFF').text_content()` throws an error. \uFEFF is not technically classified as whitespace, so the `nonspace` function which attempts to prevent processing empty documents doesn't catch it.

Solution
--------
Instead of the `isspace` method, use a regex which matches on all whitespace as well as \uFEFF.

To be applied on stable versions while the origins of stray ZWNBSPs get tackled on master.

opw-3957259

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
seems good to me

it would make sense to me to have the `\uFEFF` removed when the edition is finished, but maybe it's not possible technically (the linked commit already seems to say that we try to removed the orphaned `\uFEFF` that we can in sanitization)

note: performance wise the regex is a little slower than isspace (140ns vs 20ns in my tests if it fails on first character, but similar speed for big string with only space)  but I don't see any other solution and it doesn't seem too much

**Comment by lvsz** [comment]
> it would make sense to me to have the `\uFEFF` removed when the edition is finished

This should indeed be the case, but we haven't been able to figure out where or why exactly they're not getting removed, just that several clients are getting the same traceback. @odoo/rd-framework-editor is aware of the issue, and looking into an approach from their side.

> note: performance wise the regex is a little slower than isspace (140ns vs 20ns in my tests if it fails on first character, but similar speed for big string with only space) but I don't see any other solution and it doesn't seem too much

True, it's hard to beat [a switch case in C](https://github.com/python/cpython/blob/87cedaa5c857dc615b9f618b020414187fc1c966/Objects/unicodetype_db.h#L6374-L6412)

**Comment by dmo-odoo** [comment]
Basically, any call to `historyStep` after calling `cleanForSave` for the last time but before actually saving is susceptible to trigger that traceback. I recently tackled two of those cases with this [commit](https://github.com/odoo/odoo/pull/165678/commits/3929d9e22352dc09ec8df0464110365369a7d58b) and @deso-odoo is currently working on [that one](https://github.com/odoo/odoo/pull/167365#discussion_r1645460362). Unfortunately, we can't really know if there are other cases in practice, though we can realistically guess that there are.

The thing is, those lines of code were already wrong before, but the bug was invisible because it did not trigger a traceback back then. It does since we added the `\ufeff` mechanism. With your change, we will never find the other cases where the code has been wrong for years. Because of that, I suggest we merge your PR because we don't want customers to face tracebacks, but we should not merge it in saas-17.3 onward so that we can still be on the lookout for other cases that we still need to fix.

**Comment by lvsz** [comment]
@robodoo up to saas-17.2

File: odoo/tools/translate.py
Comment (rco-odoo) [inline odoo/tools/translate.py:174]: I personally find the variable name a bit awkward.  Why not simply `space_pattern`?

**Comment by rco-odoo** [comment]
> The thing is, those lines of code were already wrong before, but the bug was invisible because it did not trigger a traceback back then. It does since we added the `\ufeff` mechanism. With your change, we will never find the other cases where the code has been wrong for years. Because of that, I suggest we merge your PR because we don't want customers to face tracebacks, but we should not merge it in saas-17.3 onward so that we can still be on the lookout for other cases that we still need to fix.

@dmo-odoo what's the point of not merging in saas-17.3, since we will not deploy that branch?  It will never be tested for real...

File: odoo/tools/translate.py
Comment (lvsz) [reply odoo/tools/translate.py]: thought I'd name it similarly to the `isspace` method it replaces, but `space_pattern` works for me too

**Comment by dmo-odoo** [comment]
@rco-odoo Unfortunately, I don't remember what I had in mind when I mentioned saas-17.3 specifically. We can fw-port up to saas-17.3 instead of saas-17.2 if you prefer, but it won't matter if saas-17.3 is not going to be deployed anyway... ^^'

**Comment by lvsz** [comment]
@robodoo up to saas-17.3

**Review by rco-odoo — APPROVED** [review]
@robodoo r+

**Comment by xO-Tx** [comment]
Hello @dmo-odoo, apparently we still have DBs on `17.4` (odoo.com & new customer databases) where some views (saved with `/ufeff` in the arch) still trigger the same traceback when edited (because of another not-yet-fixed issue that prevents removing the `/ufeff` correctly...).

Since these views cannot be fixed in the backend (the traceback occurs on `field > write()` :grimacing:), a possible solution is to fw-port this fix to `17.4+` to unblock the situation until the issues allowing the `/ufeff` in the views are fixed :thinking:? What do you think?

cc @adwid
cc @JKE-be

**Comment by dmo-odoo** [comment]
@xO-Tx I think @deso-odoo is looking at it but if he can't find a proper fix soon then we'll have to forward port this PR to saas-17.4+, yes.

**Comment by adwid** [comment]
@dmo-odoo @deso-odoo 

Ola

Have you found anything ? 

It would be great to do something this week because we still have a lot of DBs impacted by the issue :(

**Pull request closed without merging**
